File: schema/struct.go
package schema

import (
	"bytes"
	"encoding/json"
	"runtime"
	"unsafe"

	"github.com/devsisters/go-dyncapnp/schema/proto"
)

func newStruct(ptr unsafe.Pointer) *Struct {
	return &Struct{
		Schema: NewWithFreer(ptr, releaseStructSchema),
	}
}

type Struct struct {
	*Schema
}

func (s *Struct) Type() *Type {
	return newType(mustPtr(typeFromStructSchema(s.ptr)))
}

func (s *Struct) Fields() []*StructField {
	list := structGetFields(s.ptr)
	res := make([]*StructField, len(list))
	for i, ptr := range list {
		res[i] = newStructField(ptr)
	}
	return res
}

func (s *Struct) UnionFields() []*StructField {
	list := structGetUnionFields(s.ptr)
	res := make([]*StructField, len(list))
	for i, ptr := range list {
		res[i] = newStructField(ptr)
	}
	return res
}

func (s *Struct) NonUnionFields() []*StructField {
	list := structGetNonUnionFields(s.ptr)
	res := make([]*StructField, len(list))
	for i, ptr := range list {
		res[i] = newStructField(ptr)
	}
	return res
}

func (s *Struct) Field(name string) (*StructField, bool) {
	fieldPtr, err := structFindFieldByName(s.ptr, name)
	if err != nil {
		panic(err)
	}
	if fieldPtr == nil {
		return nil, false
	}
	return newStructField(fieldPtr), true
}

func (s *Struct) Encode(json []byte) ([]byte, error) {
	return structJsonToBinary(s.ptr, json)
}

func (s *Struct) EncodePacked(json []byte) ([]byte, error) {
	return structJsonToPacked(s.ptr, json)
}

func (s *Struct) Decode(bin []byte) ([]byte, error) {
	return structBinaryToJson(s.ptr, bin)
}

func (s *Struct) DecodePacked(bin []byte) ([]byte, error) {
	return structPackedToJson(s.ptr, bin)
}

func newStructField(ptr unsafe.Pointer) *StructField {
	s := &StructField{
		ptr: ptr,
	}
	runtime.SetFinalizer(s, (*StructField).release)
	return s
}

type StructField struct {
	ptr unsafe.Pointer

	noCopy noCopy
}

func (f *StructField) Proto() (proto.Field, error) {
	b, err := structFieldToJson(f.ptr)
	if err != nil {
		return nil, err
	}
	var m proto.Field
	dec := json.NewDecoder(bytes.NewBuffer(b))
	dec.UseNumber()
	if err := dec.Decode(&m); err != nil {
		return nil, err
	}
	return m, nil
}

func (f *StructField) Parent() *Struct {
	return newStruct(structFieldGetContainingStruct(f.ptr))
}

func (f *StructField) Index() int {
	return structFieldGetIndex(f.ptr)
}

func (f *StructField) Type() *Type {
	return newType(structFieldGetType(f.ptr))
}

func (f *StructField) release() {
	releaseStructSchemaField(f.ptr)
	f.ptr = nil
	runtime.SetFinalizer(f, nil)
}

File: schema/struct.cpp
#include <cstdlib>
#include <exception>
#include <capnp/schema.h>
#include "schema.h"

template <typename T>
pointerList listToPointers(T orig) {
	auto size = orig.size();
	auto list = static_cast<void**>(malloc(sizeof(void*) * size));
	for (int i = 0; i < size; i++) {
		auto item = new decltype(orig[i]);
		*item = orig[i];
		list[i] = static_cast<void*>(item);
	}
	return {list, size};
}

pointerList structGetFields(void* structSchemaPtr) {
	auto schema = static_cast<capnp::StructSchema*>(structSchemaPtr);
	return listToPointers(schema->getFields());
}

pointerList structGetUnionFields(void* structSchemaPtr) {
	auto schema = static_cast<capnp::StructSchema*>(structSchemaPtr);
	return listToPointers(schema->getUnionFields());
}

pointerList structGetNonUnionFields(void* structSchemaPtr) {
	auto schema = static_cast<capnp::StructSchema*>(structSchemaPtr);
	return listToPointers(schema->getNonUnionFields());
}

pointer_result structFindFieldByName(void* structSchemaPtr, const char* name) {
	try {
		auto schema = static_cast<capnp::StructSchema*>(structSchemaPtr);
		KJ_IF_MAYBE(ptr, schema->findFieldByName(name)) {
			auto field = new capnp::StructSchema::Field;
			*field = *ptr;
			return {static_cast<void*>(field), nullptr};
		} else {
			return {nullptr, nullptr};
		}
	} catch(const std::exception &e) {
		return {nullptr, strdup(e.what())};
	}
}

void releaseStructSchema(void* structSchemaPtr) {
	auto schema = static_cast<capnp::StructSchema*>(structSchemaPtr);
	delete schema;
}

void* structFieldGetContainingStruct(void* structFieldPtr) {
	auto field = static_cast<capnp::StructSchema::Field*>(structFieldPtr);
	auto schema = new capnp::StructSchema;
	*schema = field->getContainingStruct();
	return static_cast<void*>(schema);
}

int structFieldGetIndex(void* structFieldPtr) {
	auto field = static_cast<capnp::StructSchema::Field*>(structFieldPtr);
	return int(field->getIndex());
}

void* structFieldGetType(void* structFieldPtr) {
	auto field = static_cast<capnp::StructSchema::Field*>(structFieldPtr);
	auto type = new capnp::Type;
	*type = field->getType();
	return static_cast<void*>(type);
}

void releaseStructSchemaField(void* structFieldPtr) {
	auto field = static_cast<capnp::StructSchema::Field*>(structFieldPtr);
	delete field;
}
